src/BlockTypes/ProductImageGallery.php

```diff
@@ -28,39 +28,142 @@ protected function get_block_type_uses_context() {
+	 * @param \WP_Post    $post    Post instance.
+	 */
+	private function sale_flash( $product, $post ) {
+		if ( ! $product->is_on_sale() ) {
+			return '';
+		}
+
+		ob_start();
+		echo sprintf( '<span class="onsale">%1$s</span>', esc_html__( 'Sale!', 'woo-gutenberg-products-block' ) );
+		return ob_get_clean();
+	}
+
+	/**
+	 * Based on the single-product/product-image.php template from the core of WooCommerce.
+	 *
+	 * @param \WC_Product $product Product instance.
+	 * @param \WP_Post    $post    Post instance.
+	 *
+	 * @return string
+	 */
+	private function product_image( $product, $post ) {
+		/**
+		 * The woocommerce_product_thumbnails_columns filter.
+		 *
+		 * Allow users to modify the number of columns displayed in the product gallery.
+		 *
+		 * @since TBD
+		 */
+		$columns           = apply_filters( 'woocommerce_product_thumbnails_columns', 4 );
+		$post_thumbnail_id = $product->get_image_id();
+		/**
+		 * The woocommerce_single_product_image_gallery_classes filter.
+		 *
+		 * Allow users to modify the class names for the product image gallery.
+		 *
+		 * @since TBD
+		 */
+		$wrapper_classes = apply_filters(
+			'woocommerce_single_product_image_gallery_classes',
+			array(
+				'woocommerce-product-gallery',
+				'woocommerce-product-gallery--' . ( $post_thumbnail_id ? 'with-images' : 'without-images' ),
+				'woocommerce-product-gallery--columns-' . absint( $columns ),
+				'images',
+			)
+		);
+
+		if ( $post_thumbnail_id ) {
+			$product_image_gallery_placeholder = wc_get_gallery_image_html( $post_thumbnail_id, true );
+		} else {
+			$product_image_gallery_placeholder = sprintf(
+				'<div class="woocommerce-product-gallery__image--placeholder"><img src="%1$s" alt="%2$s" class="wp-post-image" /></div></div>',
+				esc_url( wc_placeholder_img_src( 'woocommerce_single' ) ),
+				esc_html__( 'Awaiting product image', 'woo-gutenberg-products-block' )
+			);
+		}
+		/**
+		 * The woocommerce_single_product_image_thumbnail_html filter.
+		 *
+		 * Allow users to modify the product gallery thumbnail html.
+		 *
+		 * @since TBD
+		 */
+		$product_image_gallery_placeholder = apply_filters( 'woocommerce_single_product_image_thumbnail_html', $product_image_gallery_placeholder, $post_thumbnail_id );
+
+		return sprintf(
+			'<div class="%1$s" data-columns="%2$s"><div class="woocommerce-product-gallery__wrapper">%3$s %4$s</div></div>',
+			esc_attr( implode( ' ', array_map( 'sanitize_html_class', $wrapper_classes ) ) ),
+			esc_attr( $columns ),
+			$product_image_gallery_placeholder,
+			$this->product_thumbnails( $product )
+		);
+	}
+
+	/**
+	 * Based on the single-product/product-thumbnails.php template from the core of WooCommerce.
+	 *
+	 * @param  \WC_Product $product Product instance.
+	 *
+	 * @return false|string
+	 */
+	private function product_thumbnails( $product ) {
+		$attachment_ids = $product->get_gallery_image_ids();
+		if ( ! $attachment_ids || ! $product->get_image_id() ) {
+			return '';
+		}
+
+		ob_start();
```

Comment:
Output buffer isn't needed here, just save the output from the loop to a variable and then return it.